### PR TITLE
Grok Prepper Configuration and Boilerplate

### PR DESCRIPTION
**Pull request by graytaylor0**
Signed-off-by: graytaylor0 <33740195+graytaylor0@users.noreply.github.com>

### Description

- includes reading the configuration for Grok Prepper, which includes adding PluginSetting validation for Map<String, String>, List<String>, and Map<String, List<String>>.
-  includes boilerplate for Grok Prepper implementation of AbstractPrepper
 
### Issues Resolved
https://github.com/opensearch-project/data-prepper/issues/301
 
### Check List
- [ X ] New functionality includes testing.
  - [ X ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ X ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/grok-prepper/build.gradle [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+plugins {
+    id 'java'
+}
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    implementation "com.fasterxml.jackson.core:jackson-databind:2.12.4"
```
Comment (laneholloway) [inline data-prepper-plugins/grok-prepper/build.gradle:20]: Don't we have this already defined higher up in the hierarchy?

File: data-prepper-plugins/grok-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepperTests.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.CoreMatchers.equalTo;
+
+import java.util.Collections;
+import java.util.List;
+
+public class GrokPrepperTests {
```
Comment (laneholloway) [inline data-prepper-plugins/grok-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepperTests.java:26]: We are going to be adding more tests in here, right? ;)

**Review by laneholloway — DISMISSED** [review]

File: data-prepper-plugins/grok-prepper/build.gradle (hunk 1 above)
Comment (graytaylor0) [reply data-prepper-plugins/grok-prepper/build.gradle:20]: We actually don't. Could add this as a dependency to data-prepper-plugins but right now each individual plugin adds this if it is used.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java [hunk 3]
```diff
@@ -136,6 +137,87 @@ public String getStringOrDefault(final String attribute, final String defaultVal
...
+    @SuppressWarnings("unchecked")
+    public List<String> getStringListOrDefault(final String attribute, final List<String> defaultValue) {
+        Object object = getAttributeOrDefault(attribute, defaultValue);
+        if (object == null) {
+            return null;
+        } else if (object instanceof List) {
+            ((List<?>) object).stream().filter(o -> !(o instanceof String)).forEach(o -> {
+                throw new IllegalArgumentException(String.format(UNEXPECTED_ATTRIBUTE_TYPE_MSG, object.getClass(), attribute));
+            });
+            return (List<String>) object;
```
Comment (cmanning09) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:158]: This can be simplified:

```java
    ((List<?>) object).).forEach(o -> {
        if (!(o instanceof String)) {
                    throw new IllegalArgumentException(String.format(UNEXPECTED_ATTRIBUTE_TYPE_MSG, object.getClass(), attribute));
                }
```

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java [hunk 4]
```diff
@@ -136,6 +137,87 @@ public String getStringOrDefault(final String attribute, final String defaultVal
...
+        Object object = getAttributeOrDefault(attribute, defaultValue);
+        if (object == null) {
+            return null;
+        } else if (object instanceof Map) {
+            ((Map<?, ?>) object).forEach((key, value) -> {
+                if (key instanceof String) {
+                    if (value instanceof List){
+                        ((List<?>) value).stream().filter(o -> !(o instanceof String)).forEach(o -> {
+                            throw new IllegalArgumentException(String.format(UNEXPECTED_ATTRIBUTE_TYPE_MSG, object.getClass(), attribute));
+                        });
```
Comment (cmanning09) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:208]: See other comment above about improving this check

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java [hunk 5]
```diff
@@ -136,6 +137,87 @@ public String getStringOrDefault(final String attribute, final String defaultVal
...
+    /**
+     * Returns the value of the specified List<String>, or {@code defaultValue} if this settings contains no value for
+     * the attribute.
+     *
+     * @param attribute    name of the attribute
+     * @param defaultValue default value for the setting
+     * @return the value of the specified attribute, or {@code defaultValue} if this settings contains no value for
+     * the attribute
+     */
+    @SuppressWarnings("unchecked")
```
[the remark this answers is not in the document]
Comment (graytaylor0) [reply data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:149]: It suppresses this warning from which only comes from Intellij `Unchecked cast: 'java.lang.Object' to 'java.util.Map<java.lang.String,java.util.List<java.lang.String>>' `. It is because of the cast on the return values from Object.

File: data-prepper-plugins/grok-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepperTests.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ *  GitHub history for details.
+ */
+
+package com.amazon.dataprepper.plugins.prepper.grok;
+
+import com.amazon.dataprepper.model.configuration.PluginSetting;
+import com.amazon.dataprepper.model.record.Record;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
```
Comment (dlvenable) [inline data-prepper-plugins/grok-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepperTests.java:18]: We should start using JUnit 5 for tests.

File: data-prepper-plugins/grok-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepper.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import com.amazon.dataprepper.model.record.Record;
+import com.amazon.dataprepper.model.prepper.AbstractPrepper;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+
+
+@DataPrepperPlugin(name = "grok_prepper", type = PluginType.PREPPER)
```
Comment (dlvenable) [inline data-prepper-plugins/grok-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepper.java:28]: We should name this `grok` so that users need only type `grok` in the configuration file.

e.g.
```
prepper:
  grok:
    match: "log"
```

File: data-prepper-plugins/grok-prepper/build.gradle [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ *  GitHub history for details.
+ */
+plugins {
+    id 'java'
+}
+
+dependencies {
+    implementation project(':data-prepper-api')
+    implementation project(':data-prepper-plugins:common')
+    testImplementation project(':data-prepper-api').sourceSets.test.output
```
Comment (dlvenable) [inline data-prepper-plugins/grok-prepper/build.gradle:18]: Do we actually use this? Or was this line copied from another project?

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java [hunk 9]
```diff
@@ -136,6 +137,88 @@ public String getStringOrDefault(final String attribute, final String defaultVal
...
+    /**
+     * Returns the value of the specified List<String>, or {@code defaultValue} if this settings contains no value for
+     * the attribute.
+     *
+     * @param attribute    name of the attribute
+     * @param defaultValue default value for the setting
+     * @return the value of the specified attribute, or {@code defaultValue} if this settings contains no value for
+     * the attribute
+     */
+    public List<String> getStringListOrDefault(final String attribute, final List<String> defaultValue) {
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:149]: This method and the other methods can probably be made somewhat more generic.

```
public <T> List<T> getTypedListOrDefault(final String attribute, final Class<T> type, final List<T> defaultValue)
```

You can replace your `instanceof` with `isAssignableFrom`.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java [hunk 10]
```diff
@@ -136,6 +137,88 @@ public String getStringOrDefault(final String attribute, final String defaultVal
...
+     * @param attribute    name of the attribute
+     * @param defaultValue default value for the setting
+     * @return the value of the specified attribute, or {@code defaultValue} if this settings contains no value for
+     * the attribute
+     */
+    public Map<String, String> getStringMapOrDefault(final String attribute, final Map<String, String> defaultValue) {
+        Object object = getAttributeOrDefault(attribute, defaultValue);
+        if (object == null) {
+            return null;
+        } else if (object instanceof Map) {
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:177]: This code can be cleaner by throwing first.

I'd also create a variable just to avoid operating immediately after a cast, but I think that is of less importance.

```
if ( !(object instanceof Map))
  throw new IllegalArgumentException(...)

Map<?, ?> map = object;
map.forEach( ...
```

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java [hunk 11]
```diff
@@ -136,6 +137,88 @@ public String getStringOrDefault(final String attribute, final String defaultVal
...
+     * @param attribute    name of the attribute
+     * @param defaultValue default value for the setting
+     * @return the value of the specified attribute, or {@code defaultValue} if this settings contains no value for
+     * the attribute
+     */
+    public Map<String, List<String>> getStringListMapOrDefault(final String attribute, final Map<String, List<String>> defaultValue) {
+        Object object = getAttributeOrDefault(attribute, defaultValue);
+        if (object == null) {
+            return null;
+        } else if (object instanceof Map) {
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:201]: We should try to avoid this nesting by throwing immediately. See my note above.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/grok-prepper/build.gradle (hunk 1 above)
Comment (dlvenable) [reply data-prepper-plugins/grok-prepper/build.gradle]: It is not a dependency, but you can inherit the version number. So, you should remove the version number from this line.

See: https://github.com/opensearch-project/data-prepper/blob/main/build.gradle#L74

File: data-prepper-plugins/grok-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepper.java (hunk 7 above)
Comment (graytaylor0) [reply data-prepper-plugins/grok-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepper.java:28]: Good note

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java (hunk 10 above)
Comment (graytaylor0) [reply data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:177]: This is definitely cleaner. Just to clarify you are not suggesting that this is done immediately after the getAttributeOrDefault call, but after the if statement for checking if object is null?

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java (hunk 9 above)
Comment (graytaylor0) [reply data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:149]: Had thought about this and wasn't sure if passing the Class<T> type was considered good practice. Doing this definitely scales PluginSettings better though.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java [hunk 12]
```diff
@@ -136,6 +137,88 @@ public String getStringOrDefault(final String attribute, final String defaultVal
...
+                                throw new IllegalArgumentException(String.format(UNEXPECTED_ATTRIBUTE_TYPE_MSG, object.getClass(), attribute));
+                            }
+                        });
+                    } else {
+                        throw new IllegalArgumentException(String.format(UNEXPECTED_ATTRIBUTE_TYPE_MSG, object.getClass(), attribute));
+                    }
+                } else {
+                    throw new IllegalArgumentException(String.format(UNEXPECTED_ATTRIBUTE_TYPE_MSG, object.getClass(), attribute));
+                }
+            });
```
Comment (dinujoh) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:216]: Instance check for String, List, Map is repeated in this class. Consider defining a function or java BiPredicate that takes in the object & type, and checks if they are of same type else throws exception. This code can be reused in different methods in this class.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java [hunk 13]
```diff
@@ -180,4 +254,17 @@ public Long getLongOrDefault(final String attribute, final long defaultValue) {
         throw new IllegalArgumentException(String.format(UNEXPECTED_ATTRIBUTE_TYPE_MSG, object.getClass(), attribute));
     }
 
+    private <T> void checkObjectType(String attribute, Object object, Class<T> type) {
```
Comment (dinujoh) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java:257]: nit: final for method params line 257 & 263 to be consistent with other methods

**Review by dinujoh — APPROVED** [review]

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java (hunk 10 above)
Comment (dlvenable) [reply data-prepper-api/src/main/java/com/amazon/dataprepper/model/configuration/PluginSetting.java]: Since the behavior of a `null` value is to return `null`, then it should be after the null check.

```
Object object = getAttributeOrDefault(attribute, defaultValue);
if (object == null) {
    return null;

if ( !(object instanceof Map))
  throw new IllegalArgumentException(...)

Map<?, ?> map = object;
map.forEach( ...
```

File: settings.gradle [hunk 14]
```diff
@@ -35,4 +35,5 @@ include 'data-prepper-plugins:otel-trace-group-prepper'
 include 'data-prepper-plugins:otel-trace-source'
 include 'data-prepper-plugins:peer-forwarder'
 include 'data-prepper-plugins:blocking-buffer'
+include 'data-prepper-plugins:grok-prepper'
```
Comment (dlvenable) [inline settings.gradle:38]: Why did you add this dependency? In general, our plugins should not depend on each other.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: settings.gradle (hunk 14 above)
Comment (graytaylor0) [reply settings.gradle:38]: The settings.gradle file defines all of the internal projects. All of the plugin projects are required to go in here.

Comment (dlvenable) [reply settings.gradle:38]: Yes, you are correct on this.

**Review by dlvenable — APPROVED** [review]
